### Capstone Project- Building AI Tools with Python
Data Analysis of E-commerce data using Python and AI assistance

This e-commerce sales dataset have 3 csv files -List of Orders, Order details, Sales target with following details-
- List of Orders-This dataset contains purchase information. The information includes ID, Date of Purchase and customer details
- Order Details- This dataset contains order ID, with the order price, quantity,profit, category and subcategory of product
- Sales target-This dataset contains sales target amount and date for each product category

As part of this project, we will built a workflow, combining the above three datasets, cleaning and preparing information and generating insights using visulization. We will use AI assistance for coding and debuging in Python language. Throughout the notebook, further details will be provided as comments.

### Import required libraries and create the necessary input and output folders
Prompt for the code below- You are an experienced data analyst working in Jupyter Notebook using Python. Write clean and reusable code to install and import all required libraries for a sales analytics project. Include Pandas, NumPy, Matplotlib, Seaborn, and Pathlib. Ensure the notebook follows professional analytics project standards and is portable across systems.

In [1]:
%pip install pandas seaborn pathlib Matplotlib

Note: you may need to restart the kernel to use updated packages.


### Project Folder Structure and File Handling
prompt for the code- Create a reusable project setup for a Jupyter Notebook analytics project using Python Pathlib. Define dynamic DATA_DIR and OUTPUT_DIR path objects relative to the notebook location so the project works across different systems without changing file paths manually. Automatically create the output directory if it does not exist. Define the expected CSV filenames and raise a clear and user-friendly error if any required file is missing.

In [2]:

# --------------------------------------------------
# Define project directories relative to notebook
# --------------------------------------------------

# Assumes notebook.ipynb is inside:
# sales_analysis_project/
from pathlib import Path

PROJECT_DIR = Path.cwd()

DATA_DIR = PROJECT_DIR / "data"
OUTPUT_DIR = PROJECT_DIR / "output"

# Create output directory if it does not exist
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# --------------------------------------------------
# Define expected CSV files
# --------------------------------------------------

list_of_order = DATA_DIR / "List of Orders.csv"
Orders_Details= DATA_DIR / "Order Details.csv"
Sales_target = DATA_DIR / "Sales target.csv"

EXPECTED_FILES = [
    list_of_order,
    Orders_Details,
    Sales_target
]

# --------------------------------------------------
# Check for missing files
# --------------------------------------------------

missing_files = [file.name for file in EXPECTED_FILES if not file.exists()]

if missing_files:
    raise FileNotFoundError(
        f"Missing required data files in:\n{DATA_DIR.resolve()}\n\n"
        f"Missing files:\n- " + "\n- ".join(missing_files)
    )

# --------------------------------------------------
# Success message
# --------------------------------------------------

print("All required files found.")
print(f"Project directory : {PROJECT_DIR.resolve()}")
print(f"Data directory    : {DATA_DIR.resolve()}")
print(f"Output directory  : {OUTPUT_DIR.resolve()}")

All required files found.
Project directory : /Users/mallikasrivastava/Downloads/sales_analysis_project
Data directory    : /Users/mallikasrivastava/Downloads/sales_analysis_project/data
Output directory  : /Users/mallikasrivastava/Downloads/sales_analysis_project/output


### Reading CSV Files and Initial Inspection
Prompt- Write Python code to load the following CSV files into Pandas DataFrames:

* List of Orders.csv
* Order Details.csv
* Sales target.csv

After loading:

* print dataset shapes,
* display top 10 rows,
* display bottom 10 rows,
Use clear notebook formatting and comments.

In [3]:
import pandas as pd

# --------------------------------------------------
# Read CSV files
# --------------------------------------------------

list_of_order_df = pd.read_csv(list_of_order)
orders_details_df = pd.read_csv(Orders_Details)
sales_target_df = pd.read_csv(Sales_target)

# --------------------------------------------------
# Store datasets in dictionary
# --------------------------------------------------

datasets = {
    "List of Orders": list_of_order_df,
    "Order Details": orders_details_df,
    "Sales Target": sales_target_df
}

# --------------------------------------------------
# Print top 10 and bottom 10 rows
# --------------------------------------------------

for name, df in datasets.items():

    print("=" * 70)
    print(f"{name} Dataset")
    print("=" * 70)

    print("\nTop 10 Rows:")
    display(df.head(10))

    print("\nBottom 10 Rows:")
    display(df.tail(10))

    print(f"\nDataset Shape: {df.shape}")

    print("\n\n")


List of Orders Dataset

Top 10 Rows:


,Order ID,Order Date,CustomerName,State,City
0,B-25601,01-04-2018,Bharat,Gujarat,Ahmedabad
1,B-25602,01-04-2018,Pearl,Maharashtra,Pune
2,B-25603,03-04-2018,Jahan,Madhya Pradesh,Bhopal
3,B-25604,03-04-2018,Divsha,Rajasthan,Jaipur
4,B-25605,05-04-2018,Kasheen,West Bengal,Kolkata
5,B-25606,06-04-2018,Hazel,Karnataka,Bangalore
6,B-25607,06-04-2018,Sonakshi,Jammu and Kashmir,Kashmir
7,B-25608,08-04-2018,Aarushi,Tamil Nadu,Chennai
8,B-25609,09-04-2018,Jitesh,Uttar Pradesh,Lucknow
9,B-25610,09-04-2018,Yogesh,Bihar,Patna



Bottom 10 Rows:


,Order ID,Order Date,CustomerName,State,City
550,NaN,NaN,NaN,NaN,NaN
551,NaN,NaN,NaN,NaN,NaN
552,NaN,NaN,NaN,NaN,NaN
553,NaN,NaN,NaN,NaN,NaN
554,NaN,NaN,NaN,NaN,NaN
555,NaN,NaN,NaN,NaN,NaN
556,NaN,NaN,NaN,NaN,NaN
557,NaN,NaN,NaN,NaN,NaN
558,NaN,NaN,NaN,NaN,NaN
559,NaN,NaN,NaN,NaN,NaN



Dataset Shape: (560, 5)



Order Details Dataset

Top 10 Rows:


,Order ID,Amount,Profit,Quantity,Category,Sub-Category
0,B-25601,1275.0,-1148.0,7,Furniture,Bookcases
1,B-25601,66.0,-12.0,5,Clothing,Stole
2,B-25601,8.0,-2.0,3,Clothing,Hankerchief
3,B-25601,80.0,-56.0,4,Electronics,Electronic Games
4,B-25602,168.0,-111.0,2,Electronics,Phones
5,B-25602,424.0,-272.0,5,Electronics,Phones
6,B-25602,2617.0,1151.0,4,Electronics,Phones
7,B-25602,561.0,212.0,3,Clothing,Saree
8,B-25602,119.0,-5.0,8,Clothing,Saree
9,B-25603,1355.0,-60.0,5,Clothing,Trousers



Bottom 10 Rows:


,Order ID,Amount,Profit,Quantity,Category,Sub-Category
1490,B-26098,409.0,86.0,3,Clothing,Saree
1491,B-26098,59.0,15.0,2,Clothing,T-shirt
1492,B-26098,46.0,14.0,5,Clothing,Skirt
1493,B-26099,9.0,3.0,1,Clothing,Skirt
1494,B-26099,207.0,37.0,4,Clothing,Hankerchief
1495,B-26099,835.0,267.0,5,Electronics,Phones
1496,B-26099,2366.0,552.0,5,Clothing,Trousers
1497,B-26100,828.0,230.0,2,Furniture,Chairs
1498,B-26100,34.0,10.0,2,Clothing,T-shirt
1499,B-26100,72.0,16.0,2,Clothing,Shirt



Dataset Shape: (1500, 6)



Sales Target Dataset

Top 10 Rows:


,Month of Order Date,Category,Target
0,Apr-18,Furniture,10400.0
1,May-18,Furniture,10500.0
2,Jun-18,Furniture,10600.0
3,Jul-18,Furniture,10800.0
4,Aug-18,Furniture,10900.0
5,Sep-18,Furniture,11000.0
6,Oct-18,Furniture,11100.0
7,Nov-18,Furniture,11300.0
8,Dec-18,Furniture,11400.0
9,Jan-19,Furniture,11500.0



Bottom 10 Rows:


,Month of Order Date,Category,Target
26,Jun-18,Electronics,9000.0
27,Jul-18,Electronics,9000.0
28,Aug-18,Electronics,9000.0
29,Sep-18,Electronics,9000.0
30,Oct-18,Electronics,9000.0
31,Nov-18,Electronics,9000.0
32,Dec-18,Electronics,9000.0
33,Jan-19,Electronics,16000.0
34,Feb-19,Electronics,16000.0
35,Mar-19,Electronics,16000.0



Dataset Shape: (36, 3)





### Dataset Audit Function
Prompt- Write a reusable audit function in Python using Pandas that generates a compact data quality report for the three DataFrame. The report should include:

* row and column counts,
* column datatypes,
* missing value counts,
* duplicate row counts,
* and basic structural validation.

In [4]:
# ==========================================================
# COMPACT DATA AUDIT FUNCTION
# ==========================================================

def audit_df(df, name="DataFrame"):
    """
    Prints a compact data quality summary:
    - Shape (rows, columns)
    - Column data types
    - Missing value counts
    """

    print("=" * 70)
    print(f"AUDIT REPORT: {name}")
    print("=" * 70)

    # Row & column count
    print(f"\nShape: {df.shape[0]} rows × {df.shape[1]} columns\n")

    # Data types
    print("Column Data Types:")
    print(df.dtypes)
    
    print("\nMissing Values:")
    missing = df.isnull().sum()
    print(missing[missing > 0] if missing.sum() > 0 else "No missing values 🎉")

    print("\n")

In [5]:
audit_df(list_of_order_df, "List of Orders")
audit_df(orders_details_df, "Order Details")
audit_df(sales_target_df, "Sales Target")

AUDIT REPORT: List of Orders

Shape: 560 rows × 5 columns

Column Data Types:
Order ID        object
Order Date      object
CustomerName    object
State           object
City            object
dtype: object

Missing Values:
Order ID        60
Order Date      60
CustomerName    60
State           60
City            60
dtype: int64


AUDIT REPORT: Order Details

Shape: 1500 rows × 6 columns

Column Data Types:
Order ID         object
Amount          float64
Profit          float64
Quantity          int64
Category         object
Sub-Category     object
dtype: object

Missing Values:
No missing values 🎉


AUDIT REPORT: Sales Target

Shape: 36 rows × 3 columns

Column Data Types:
Month of Order Date     object
Category                object
Target                 float64
dtype: object

Missing Values:
No missing values 🎉




### Data quality Observation and Cleaning-
we see null rows in list of orders, we see trailing spaces,different date formats and negative numbers in profit columns in order details table.
Prompt- wrote a code for data cleaning by removing fully empty rows,strip spaces from the possible joining keys,convert all dates to standard format in the three dataframes. Create 'month' field in the List of orders dataframe using Order Date  field to be used as joining key with Sales Target data. Suggest any other feature engineering requirements for performing merge and EDA in the next steps.

In [6]:
# ==========================================================
# DATA QUALITY CLEANING PIPELINE
# For: List of Orders, Order Details, Sales Target
# ==========================================================

# -----------------------------
# 1. REMOVE FULLY EMPTY ROWS
# Issue: trailing blank rows in List of Orders
# -----------------------------
list_of_order_df = list_of_order_df.dropna(how="all")

# -----------------------------
# 2. STANDARDIZE JOIN KEYS
# Issue: hidden spaces / inconsistent types in Order ID
# -----------------------------
list_of_order_df["Order ID"] = list_of_order_df["Order ID"].astype(str).str.strip()
orders_details_df["Order ID"] = orders_details_df["Order ID"].astype(str).str.strip()

# -----------------------------
# 3. CONVERT ORDER DATE (LIST OF ORDERS)
# Issue: date stored as string (DD-MM-YYYY)
# -----------------------------
list_of_order_df["Order Date"] = pd.to_datetime(
    list_of_order_df["Order Date"],
    format="%d-%m-%Y",
    errors="coerce"
)

# -----------------------------
# 4. CONVERT SALES TARGET MONTH COLUMN
# Issue: month stored as string (Apr-18)
# -----------------------------
sales_target_df["Month of Order Date"] = pd.to_datetime(
    sales_target_df["Month of Order Date"],
    format="%b-%y",
    errors="coerce"
)

# -----------------------------
# 5. CONVERT NUMERIC COLUMNS SAFELY
# Issue: numeric columns may be object type or corrupted
# -----------------------------
for col in ["Amount", "Profit", "Quantity"]:
    orders_details_df[col] = pd.to_numeric(
        orders_details_df[col],
        errors="coerce"
    )

sales_target_df["Target"] = pd.to_numeric(
    sales_target_df["Target"],
    errors="coerce"
)

# -----------------------------
# 6. CONVERT CATEGORICAL COLUMNS
# Issue: repeated string columns optimized to category type
# -----------------------------
cat_cols_orders = ["CustomerName", "State", "City"]
for col in cat_cols_orders:
    list_of_order_df[col] = (
        list_of_order_df[col]
        .astype(str)
        .str.strip()
        .astype("category")
    )

cat_cols_details = ["Category", "Sub-Category"]
for col in cat_cols_details:
    orders_details_df[col] = (
        orders_details_df[col]
        .astype(str)
        .str.strip()
        .astype("category")
    )

sales_target_df["Category"] = (
    sales_target_df["Category"]
    .astype(str)
    .str.strip()
    .astype("category")
)

# -----------------------------
# 7. HANDLE MISSING VALUES SAFELY
# Issue: NaNs after conversions (especially dates/numerics)
# -----------------------------
# Option: drop or keep depending on analysis need
list_of_order_df = list_of_order_df.dropna(subset=["Order ID", "Order Date"])
orders_details_df = orders_details_df.dropna(subset=["Order ID"])

# -----------------------------
# 8. CREATE TIME FEATURES (FOR FUTURE EDA)
# -----------------------------
list_of_order_df["Order Month"] = list_of_order_df["Order Date"].dt.to_period("M").dt.to_timestamp()

# -----------------------------
# 9. VALIDATE DATA QUALITY
# -----------------------------
print("DATA QUALITY CHECK SUMMARY")
print("=" * 40)

print("\nMissing values per dataset:")
print("Orders:\n", list_of_order_df.isnull().sum())
print("\nOrder Details:\n", orders_details_df.isnull().sum())
print("\nSales Target:\n", sales_target_df.isnull().sum())

print("\nShapes after cleaning:")
print("Orders:", list_of_order_df.shape)
print("Order Details:", orders_details_df.shape)
print("Sales Target:", sales_target_df.shape)

print("\nDuplicate Order IDs (Orders):", list_of_order_df["Order ID"].duplicated().sum())
print("Duplicate Order IDs (Details):", orders_details_df["Order ID"].duplicated().sum())

# -----------------------------
# 10. READY FOR MERGE CONFIRMATION
# -----------------------------
print("\nData cleaning completed. Ready for merging and EDA.")

DATA QUALITY CHECK SUMMARY

Missing values per dataset:
Orders:
 Order ID        0
Order Date      0
CustomerName    0
State           0
City            0
Order Month     0
dtype: int64

Order Details:
 Order ID        0
Amount          0
Profit          0
Quantity        0
Category        0
Sub-Category    0
dtype: int64

Sales Target:
 Month of Order Date    0
Category               0
Target                 0
dtype: int64

Shapes after cleaning:
Orders: (500, 6)
Order Details: (1500, 6)
Sales Target: (36, 3)

Duplicate Order IDs (Orders): 0
Duplicate Order IDs (Details): 1000

Data cleaning completed. Ready for merging and EDA.


<ipython-input-6-951dbc110196>:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  list_of_order_df["Order ID"] = list_of_order_df["Order ID"].astype(str).str.strip()
<ipython-input-6-951dbc110196>:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  list_of_order_df["Order Date"] = pd.to_datetime(
<ipython-input-6-951dbc110196>:60: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentat

### Advance Data Cleaning 
Prompt- Ensure joining keys are unique and not missing. This will help in avoiding any data loss/rows drops during merge and keep the total consistent.
Prompt- scenarios excluding where entire row is empty, look for the possible unique keys or joining keys and if that value is missing, replace missing or blank customer_id values with stable UNKNOWN_ IDs, and print how many IDs were filled. Ensure to follow the join structure of- 
List of Orders ↔ Order Details → Order ID
Sales Target ↔ Category + Month
So focus on:
Order ID
Category
Month of Order Date

In [7]:
# ==========================================================
# FUNCTION: FIX MISSING JOIN KEYS SAFELY
# ==========================================================

def fix_missing_keys(df, key_cols, df_name="DataFrame"):
    """
    Replaces missing/blank join keys with stable UNKNOWN_ IDs
    and prints how many values were fixed.
    """

    print("=" * 70)
    print(f"KEY CLEANING REPORT: {df_name}")
    print("=" * 70)

    for col in key_cols:

        if col not in df.columns:
            print(f"[SKIP] Column '{col}' not found")
            continue

        # Standardize blanks → NaN
        df[col] = df[col].astype(str).str.strip()
        df[col] = df[col].replace(["", "nan", "None", "NaN"], pd.NA)

        # Count missing before fill
        missing_count = df[col].isna().sum()

        # Fill missing with stable IDs
        if missing_count > 0:
            if col == "Order ID":
                df[col] = df[col].fillna("UNKNOWN_ORDER_ID")
            elif col == "Category":
                df[col] = df[col].fillna("UNKNOWN_CATEGORY")
            elif col == "Month of Order Date":
                df[col] = df[col].fillna("UNKNOWN_MONTH")
            else:
                df[col] = df[col].fillna(f"UNKNOWN_{col.upper()}")

        print(f"{col}: filled {missing_count} missing values")

    print("\n")
    return df

In [8]:

# -----------------------------
# 1. List of Orders
# -----------------------------
list_of_order_df = fix_missing_keys(
    list_of_order_df,
    key_cols=["Order ID"],
    df_name="List of Orders"
)

# -----------------------------
# 2. Order Details
# -----------------------------
orders_details_df = fix_missing_keys(
    orders_details_df,
    key_cols=["Order ID", "Category"],
    df_name="Order Details"
)

# -----------------------------
# 3. Sales Target
# -----------------------------
sales_target_df = fix_missing_keys(
    sales_target_df,
    key_cols=["Category", "Month of Order Date"],
    df_name="Sales Target"
)

KEY CLEANING REPORT: List of Orders
Order ID: filled 0 missing values


KEY CLEANING REPORT: Order Details
Order ID: filled 0 missing values
Category: filled 0 missing values


KEY CLEANING REPORT: Sales Target
Category: filled 0 missing values
Month of Order Date: filled 0 missing values




### Advance Data Cleaning-check for duplicates in Order IDs fields to ensure data integrity of unique keys.
Prompt- Create reusable Python functions to:

* detect duplicate rows,
* distinguish between business-valid duplicates and true data quality issues,
* remove invalid duplicate records,
* and provide summary reporting.

Explain why duplicate Order IDs may be expected in transactional line-item datasets and how to correctly interpret them during analysis.


In [9]:
print("Duplicate Order IDs (Orders):",
      list_of_order_df["Order ID"].duplicated().sum())

print("Duplicate Order IDs (Details):",
      orders_details_df["Order ID"].duplicated().sum())

Duplicate Order IDs (Orders): 0
Duplicate Order IDs (Details): 1000


analysis- List of Orders = HEADER TABLE
Each row = one order- good
2. Order Details = LINE ITEM TABLE
Each row = one product inside an order
Example:
Order ID	Category	Amount
B-25601	Furniture	1275
B-25601	Clothing	66
B-25601	Electronics	80
so the behaviour is expected.

Orders = dimension table
Order Details = fact table (transaction lines)
So duplicates in Order Details = GOOD DATA DESIGN

### Check distribution of order id in Transcation tables for order size and any anomalies

In [10]:
#Check distribution of order size
orders_details_df["Order ID"].value_counts().describe()

count    500.000000
mean       3.000000
std        2.478768
min        1.000000
25%        1.000000
50%        2.000000
75%        4.000000
max       12.000000
Name: Order ID, dtype: float64

### Final Data Quality Snapshot

In [11]:
#FINAL DATA QUALITY SNAPSHOT (before merge)
def quick_snapshot(df, name):
    print("=" * 60)
    print(f"{name} - DATA SNAPSHOT")
    print("=" * 60)
    
    print("Shape:", df.shape)
    print("\nMissing Values:")
    print(df.isnull().sum()[df.isnull().sum() > 0])
    
    print("\nDuplicates:", df.duplicated().sum())
    print("\nDtypes:\n", df.dtypes)
    print("\n")

quick_snapshot(list_of_order_df, "List of Orders")
quick_snapshot(orders_details_df, "Order Details")
quick_snapshot(sales_target_df, "Sales Target")

List of Orders - DATA SNAPSHOT
Shape: (500, 6)

Missing Values:
Series([], dtype: int64)

Duplicates: 0

Dtypes:
 Order ID                object
Order Date      datetime64[ns]
CustomerName          category
State                 category
City                  category
Order Month     datetime64[ns]
dtype: object


Order Details - DATA SNAPSHOT
Shape: (1500, 6)

Missing Values:
Series([], dtype: int64)

Duplicates: 0

Dtypes:
 Order ID          object
Amount           float64
Profit           float64
Quantity           int64
Category          object
Sub-Category    category
dtype: object


Sales Target - DATA SNAPSHOT
Shape: (36, 3)

Missing Values:
Series([], dtype: int64)

Duplicates: 0

Dtypes:
 Month of Order Date     object
Category                object
Target                 float64
dtype: object




In [12]:
#FINAL TYPE STANDARDIZATION (merge-safe)

# -----------------------------
# Order ID standardization
# -----------------------------
list_of_order_df["Order ID"] = list_of_order_df["Order ID"].astype(str).str.strip()
orders_details_df["Order ID"] = orders_details_df["Order ID"].astype(str).str.strip()

# -----------------------------
# Category standardization
# -----------------------------
orders_details_df["Category"] = orders_details_df["Category"].astype(str).str.strip()
sales_target_df["Category"] = sales_target_df["Category"].astype(str).str.strip()

# -----------------------------
# Date standardization
# -----------------------------
list_of_order_df["Order Date"] = pd.to_datetime(list_of_order_df["Order Date"], errors="coerce")

sales_target_df["Month of Order Date"] = pd.to_datetime(
    sales_target_df["Month of Order Date"],
    format="%b-%y",
    errors="coerce"
)

In [13]:
##REMOVE DUPLICATES (FINAL PASS)

list_of_order_df = list_of_order_df.drop_duplicates()
orders_details_df = orders_details_df.drop_duplicates()
sales_target_df = sales_target_df.drop_duplicates()

print("Duplicates removed and confirmed clean")

Duplicates removed and confirmed clean


### Perform OUTER MERGE on three datasets and check the data integrity with audit of the final dataset
Prompt- perform outer merge on the list of orders and order details dataset with following conditions-
keeps all records= outer join
validate-one to many
flags origin using indicator=True
create a readable merge status column and help detect dropped/mismatched rows with an audit report checking all required columns, count of rows and missing/mismatches values

In [14]:

# ==========================================================
# OUTER MERGE WITH FULL AUDIT TRAIL
# ==========================================================

sales_df = pd.merge(
    list_of_order_df,
    orders_details_df,
    on="Order ID",
    how="outer",
    indicator=True,
    validate="one_to_many"   # correct relationship
)

# ----------------------------------------------------------
# Create readable merge status column
# ----------------------------------------------------------

sales_df["merge_status"] = sales_df["_merge"].map({
    "both": "Matched in both datasets",
    "left_only": "Only in Orders (missing details)",
    "right_only": "Only in Order Details (orphan rows)"
})

# ----------------------------------------------------------
# Quick summary of merge quality
# ----------------------------------------------------------

print("=" * 60)
print("MERGE QUALITY REPORT")
print("=" * 60)

print("\nRow counts by source:")
print(sales_df["_merge"].value_counts())

print("\nTotal rows after merge:", sales_df.shape[0])

print("\nMerge status breakdown:")
print(sales_df["merge_status"].value_counts())


MERGE QUALITY REPORT

Row counts by source:
both          1500
right_only       0
left_only        0
Name: _merge, dtype: int64

Total rows after merge: 1500

Merge status breakdown:
Matched in both datasets               1500
Only in Order Details (orphan rows)       0
Only in Orders (missing details)          0
Name: merge_status, dtype: int64


In [15]:
#Filter clean dataset for analysis
clean_sales_df = sales_df[sales_df["_merge"] == "both"].copy()
#Drop helper column
clean_sales_df.drop(columns=["_merge"], inplace=True)


#### MERGE WITH SALES TARGET (OUTER JOIN + AUDIT)
prompt-join the sales_df (orders + order details) dataset with the sales_target_df (Month + Category targets) using following criteria- outer join (no data loss hidden)
indicator=True
create a readable merge status column and help detect dropped/mismatched rows with an audit report checking all required columns, count of rows and missing/mismatches values

In [16]:

# ==========================================================
# MERGE WITH SALES TARGET (OUTER JOIN + AUDIT)
# ==========================================================

sales_target_merged = pd.merge(
    sales_df,
    sales_target_df,
    left_on=["Order Month", "Category"],
    right_on=["Month of Order Date", "Category"],
    how="outer",
    indicator="target_merge_flag",
    validate="many_to_many"
)

# ----------------------------------------------------------
# Create readable merge status
# ----------------------------------------------------------

sales_target_merged["merge_status"] = (
    sales_target_merged["target_merge_flag"]
    .map({
        "both": "Matched in Actual + Target",
        "left_only": "Actual only (no target match)",
        "right_only": "Target only (no actual sales)"
    })
)

# ----------------------------------------------------------
# Summary report
# ----------------------------------------------------------

print("=" * 70)
print("SALES vs TARGET MERGE QUALITY REPORT")
print("=" * 70)

print("\nRow distribution:")
print(sales_target_merged["target_merge_flag"].value_counts())

print("\nTotal rows after merge:")
print(sales_target_merged.shape[0])

print("\nStatus breakdown:")
print(sales_target_merged["merge_status"].value_counts())

SALES vs TARGET MERGE QUALITY REPORT

Row distribution:
left_only     1500
right_only      17
both             0
Name: target_merge_flag, dtype: int64

Total rows after merge:
1517

Status breakdown:
Actual only (no target match)    1500
Target only (no actual sales)      17
Matched in Actual + Target          0
Name: merge_status, dtype: int64


### Explortatory Data Analysis
prompt- Using Seaborn and Matplotlib, generate professional exploratory data analysis visualizations for:

* monthly sales trends,
* category-wise sales,
* category-wise profitability,
* and actual vs target performance.

Each chart should include:

* clear titles,
* proper axis labels,
* suitable color palettes,
* and clean formatting for executive presentation.
For each EDA visualization, write a 5–6 line management-level business summary suitable for senior leadership review. The observations should:

* explain key business trends,
* identify opportunities and risks,
* highlight operational implications,
* and support strategic decision-making.

The tone should be concise, professional, and executive-friendly rather than overly technical.

### Create the HTML version of the graphs.
prompt-Give me a python code to create a html version of the graphs in a dashboard style for easy view.

In [17]:
import os
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style("whitegrid")

# Create output directory
os.makedirs("eda_html_output", exist_ok=True)

# ==========================================================
# GRAPH 1 — MONTHLY SALES TREND
# ==========================================================

monthly_sales = (
    sales_df
    .groupby("Order Month", as_index=False)["Amount"]
    .sum()
)

plt.figure(figsize=(12,6))

sns.lineplot(
    data=monthly_sales,
    x="Order Month",
    y="Amount",
    marker="o",
    linewidth=2.5,
    color="royalblue"
)

plt.title("Monthly Sales Trend", weight="bold")
plt.xlabel("Order Month")
plt.ylabel("Total Sales")
plt.xticks(rotation=45)
plt.tight_layout()

plt.savefig("eda_html_output/monthly_sales.png", dpi=300)
plt.close()

# ==========================================================
# GRAPH 2 — CATEGORY SALES
# ==========================================================

category_sales = (
    sales_df
    .groupby("Category", as_index=False)["Amount"]
    .sum()
    .sort_values("Amount", ascending=False)
)

plt.figure(figsize=(12,6))

sns.barplot(
    data=category_sales,
    x="Category",
    y="Amount",
    palette="viridis"
)

plt.title("Category-wise Total Sales", weight="bold")
plt.tight_layout()

plt.savefig("eda_html_output/category_sales.png", dpi=300)
plt.close()

# ==========================================================
# GRAPH 3 — CATEGORY PROFITABILITY
# ==========================================================

category_profit = (
    sales_df
    .groupby("Category", as_index=False)["Profit"]
    .sum()
    .sort_values("Profit", ascending=False)
)

plt.figure(figsize=(12,6))

sns.barplot(
    data=category_profit,
    x="Category",
    y="Profit",
    palette="coolwarm"
)

plt.axhline(0, color="black")
plt.title("Category-wise Profit Analysis", weight="bold")
plt.tight_layout()

plt.savefig("eda_html_output/category_profit.png", dpi=300)
plt.close()

# ==========================================================
# GRAPH 4 — TARGET VS ACTUAL
# ==========================================================

target_vs_actual = (
    sales_target_merged
    .groupby("Category", as_index=False)
    .agg({
        "Amount": "sum",
        "Target": "sum"
    })
)

target_vs_actual = target_vs_actual.melt(
    id_vars="Category",
    value_vars=["Amount", "Target"],
    var_name="Metric",
    value_name="Value"
)

plt.figure(figsize=(12,6))

sns.barplot(
    data=target_vs_actual,
    x="Category",
    y="Value",
    hue="Metric",
    palette="Set2"
)

plt.title("Actual Sales vs Sales Target", weight="bold")
plt.tight_layout()

plt.savefig("eda_html_output/target_vs_actual.png", dpi=300)
plt.close()



In [33]:
html_content = """
<!DOCTYPE html>
<html>
<head>
<title>Sales EDA Dashboard</title>

<style>

body{
    font-family: Arial, sans-serif;
    margin:40px;
    background:#f4f6f9;
    color:#333;
}

h1{
    text-align:center;
    color:#1f4e79;
    margin-bottom:40px;
}

.card{
    background:white;
    padding:25px;
    margin-bottom:35px;
    border-radius:12px;
    box-shadow:0 2px 10px rgba(0,0,0,0.08);
}

table{
    width:100%;
    border-collapse:collapse;
    margin-top:20px;
}

th, td{
    border:1px solid #ddd;
    padding:12px;
    text-align:left;
}

th{
    background:#1f4e79;
    color:white;
}

</style>

</head>

<body>

<h1>Data Analysis of E-Commerce Data Using Python and AI Assistance</h1>

<div class="card">

<h2>Project Overview</h2>

<p style="line-height:1.8; font-size:16px;">

This capstone project focuses on analyzing e-commerce sales data using Python,
data analytics libraries, and AI-assisted prompt engineering techniques.
The project aims to perform data cleaning, exploratory data analysis (EDA),
business performance evaluation, and dashboard reporting to derive meaningful
business insights from transactional sales data.

The analysis includes sales trends, category-level revenue contribution,
profitability analysis, and target versus actual sales comparison.

</p>

</div>


<div class="card">

<h2>Dataset Summary</h2>

<table>

<tr>
<th>Dataset / Area</th>
<th>Description</th>
<th>Key Columns Used</th>
</tr>

<tr>
<td>List of Orders</td>
<td>Contains customer order-level information.</td>
<td>Order ID, Order Date, State, City</td>
</tr>

<tr>
<td>Order Details</td>
<td>Contains transactional sales details.</td>
<td>Category, Amount, Profit, Quantity</td>
</tr>

<tr>
<td>Sales Target</td>
<td>Contains monthly category-level sales targets.</td>
<td>Month, Category, Target</td>
</tr>

<tr>
<td>Data Cleaning</td>
<td>Performed preprocessing and validation checks.</td>
<td>Missing values, duplicates, formatting</td>
</tr>

<tr>
<td>EDA & Reporting</td>
<td>Business analysis and visualization reporting.</td>
<td>Sales trends, profitability, target analysis</td>
</tr>

</table>

</div>
"""

In [36]:
# ==========================================================
# CREATE HTML DASHBOARD WITH INSIGHTS
# ==========================================================

html_content = """
<!DOCTYPE html>
<html>
<head>
<title>Sales EDA Dashboard</title>

<style>

body{
    font-family: Arial, sans-serif;
    margin:40px;
    background:#f4f6f9;
    color:#333;
}

h1{
    text-align:center;
    color:#1f4e79;
    margin-bottom:40px;
}

.card{
    background:white;
    padding:25px;
    margin-bottom:35px;
    border-radius:12px;
    box-shadow:0 2px 10px rgba(0,0,0,0.08);
}

table{
    width:100%;
    border-collapse:collapse;
    margin-top:20px;
}

th, td{
    border:1px solid #ddd;
    padding:12px;
    text-align:left;
}

th{
    background:#1f4e79;
    color:white;
}

</style>

</head>

<body>

<h1>Data Analysis of E-Commerce Data Using Python and AI Assistance</h1>

<div class="card">

<h2>Project Overview</h2>

<p style="line-height:1.8; font-size:16px;">

This capstone project focuses on analyzing e-commerce sales data using Python,
data analytics libraries, and AI-assisted prompt engineering techniques.
The project aims to perform data cleaning, exploratory data analysis (EDA),
business performance evaluation, and dashboard reporting to derive meaningful
business insights from transactional sales data.

The analysis includes sales trends, category-level revenue contribution,
profitability analysis, and target versus actual sales comparison.

</p>

</div>


<div class="card">

<h2>Dataset Summary</h2>

<table>

<tr>
<th>Dataset / Area</th>
<th>Description</th>
<th>Key Columns Used</th>
</tr>

<tr>
<td>List of Orders</td>
<td>Contains customer order-level information.</td>
<td>Order ID, Order Date, State, City</td>
</tr>

<tr>
<td>Order Details</td>
<td>Contains transactional sales details.</td>
<td>Category, Amount, Profit, Quantity</td>
</tr>

<tr>
<td>Sales Target</td>
<td>Contains monthly category-level sales targets.</td>
<td>Month, Category, Target</td>
</tr>

<tr>
<td>Data Cleaning</td>
<td>Performed preprocessing and validation checks.</td>
<td>Missing values, duplicates, formatting</td>
</tr>

<tr>
<td>EDA & Reporting</td>
<td>Business analysis and visualization reporting.</td>
<td>Sales trends, profitability, target analysis</td>
</tr>

</table>

</div>

<!DOCTYPE html>
<html>
<head>
<title>Sales EDA Dashboard</title>

<style>

body{
    font-family: Arial, sans-serif;
    margin:40px;
    background:#f4f6f9;
    color:#333;
}

h1{
    text-align:center;
    color:#1f4e79;
    margin-bottom:40px;
}

.card{
    background:white;
    padding:25px;
    margin-bottom:35px;
    border-radius:12px;
    box-shadow:0 2px 10px rgba(0,0,0,0.08);
}

.card h2{
    color:#1f4e79;
    margin-bottom:20px;
}

img{
    width:100%;
    max-width:1200px;
    border-radius:8px;
}

.insights{
    margin-top:20px;
    padding:18px;
    background:#eef4fb;
    border-left:5px solid #1f4e79;
    border-radius:8px;
    line-height:1.7;
}

.insights h3{
    margin-top:0;
    color:#1f4e79;
}

</style>
</head>

<body>

<h1>Sales Performance Dashboard</h1>


<!-- ===================================================== -->
<!-- GRAPH 1 -->
<!-- ===================================================== -->

<div class="card">

<h2>Monthly Sales Trend</h2>

<img src="monthly_sales.png">

<div class="insights">
<h3>Insights</h3>

<p>
This visualization measures total sales revenue across different months,
helping identify seasonal demand patterns and overall business performance trends.
The chart shows fluctuations in monthly revenue, indicating that certain periods
contribute significantly more to sales, likely driven by promotions, festive demand,
or market purchasing cycles.
</p>

</div>

</div>


<!-- ===================================================== -->
<!-- GRAPH 2 -->
<!-- ===================================================== -->

<div class="card">

<h2>Category-wise Sales Performance</h2>

<img src="category_sales.png">

<div class="insights">
<h3>Insights</h3>

<p>
This chart compares total sales generated by each product category to evaluate
revenue contribution across business segments. The visualization highlights that
a few categories dominate overall sales performance, suggesting stronger customer
demand and higher market preference within specific product groups.
</p>

</div>

</div>


<!-- ===================================================== -->
<!-- GRAPH 3 -->
<!-- ===================================================== -->

<div class="card">

<h2>Category-wise Profitability</h2>

<img src="category_profit.png">

<div class="insights">
<h3>Insights</h3>

<p>
This analysis measures total profit generated by each product category to evaluate
financial performance beyond revenue alone. The visualization reveals that strong
sales performance does not always translate into high profitability, indicating
possible margin pressure, discounting impacts, or higher operational costs in
certain categories.
</p>

</div>

</div>


<!-- ===================================================== -->
<!-- GRAPH 4 -->
<!-- ===================================================== -->

<div class="card">

<h2>Target vs Actual Sales</h2>

<img src="target_vs_actual.png">

<div class="insights">
<h3>Insights</h3>

<p>
This visualization compares achieved sales performance against predefined business
targets across categories. The chart helps identify areas where performance exceeded
expectations as well as segments that underperformed, providing valuable insight into
execution effectiveness and opportunities for strategic improvement.
</p>

</div>

</div>

</body>
</html>
"""

# ==========================================================
# SAVE HTML FILE
# ==========================================================

with open("eda_html_output/Sales_Dashboard.html", "w", encoding="utf-8") as f:
    f.write(html_content)

print("HTML dashboard with insights created successfully.")

HTML dashboard with insights created successfully.
